### ISSUE TITLE
include more context with exceptions

### ISSUE THREAD

**Issue by reedsa** [ethereum/web3.py#3094]
RPC errors are handled in various ways depending on the format of the response. When a response is malformed in a way that raises a BadResponseFormat, the raw response is appended to the error message, making it difficult for humans to parse.

Other exceptions like ValueError could also provide better messages and information.

For this effort, consider splitting up the message from the response as separate arguments to the excpetions. Also provide more context about the errors if possible.

It may be useful to audit other exceptions in web3.py to see if they can be improved as well.

**Comment by fselmo** [comment]
related to #2798

**Comment by Wenzhi-Ding** [comment]
Yes. I have a related issue. Currently, when there is a timeout error from RPC provider, current `manager.py` returns `ValueError`, which is difficult to catch and handle. There are too many possibilities for `ValueError`. Handling a timeout is different from handling ordinary `ValueError`. I suggest to seperate this error to someting like `RPCTimeoutError`.

```
Traceback (most recent call last):
  File "/home/ubuntu/collection_2401/app/submit_txn.py", line 160, in <module>
    pick_txns()
  File "/home/ubuntu/collection_2401/app/submit_txn.py", line 73, in pick_txns
    submit_txn(txn, nonce)
  File "/home/ubuntu/collection_2401/app/submit_txn.py", line 100, in submit_txn
    txn_hash = _mint(
               ^^^^^^
  File "/home/ubuntu/collection_2401/utils/web3_write.py", line 44, in _mint
    "gasPrice": w3.eth.generate_gas_price() * 2
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/miniconda3/lib/python3.11/site-packages/web3/eth/base_eth.py", line 89, in generate_gas_price
    return self._gas_price_strategy(self.w3, transaction_params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/miniconda3/lib/python3.11/site-packages/web3/gas_strategies/time_based.py", line 232, in time_based_gas_price_strategy
    probabilities = _compute_probabilities(
                    ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/miniconda3/lib/python3.11/site-packages/eth_utils/functional.py", line 45, in inner
    return callback(fn(*args, **kwargs))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/miniconda3/lib/python3.11/site-packages/web3/gas_strategies/time_based.py", line 133, in _compute_probabilities
    sorted(
  File "/home/ubuntu/miniconda3/lib/python3.11/site-packages/web3/gas_strategies/time_based.py", line 107, in _aggregate_miner_data
    data_by_miner = groupby(0, raw_data)
                    ^^^^^^^^^^^^^^^^^^^^
  File "cytoolz/itertoolz.pyx", line 116, in cytoolz.itertoolz.groupby
  File "cytoolz/itertoolz.pyx", line 160, in cytoolz.itertoolz.groupby
  File "/home/ubuntu/miniconda3/lib/python3.11/site-packages/web3/gas_strategies/time_based.py", line 98, in _get_raw_miner_data
    block = w3.eth.get_block(block["parentHash"], full_transactions=True)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/miniconda3/lib/python3.11/site-packages/web3/eth/eth.py", line 390, in get_block
    return self._get_block(block_identifier, full_transactions)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/miniconda3/lib/python3.11/site-packages/web3/module.py", line 75, in caller
    result = w3.manager.request_blocking(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/miniconda3/lib/python3.11/site-packages/web3/manager.py", line 325, in request_blocking
    return self.formatted_response(
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/miniconda3/lib/python3.11/site-packages/web3/manager.py", line 288, in formatted_response
    raise ValueError(error)
ValueError: {'code': -32603, 'message': 'request failed or timed out'}
```

**Comment by fselmo** [comment]
@reedsa double check if the scope was a bit wider here. Otherwise this looks like it was closed by #3263 + #3282.
